Add predicate filtering to queuing handler

wallrj-cyberark opened the pull request:
- Register GenerationChangedPredicate for ACME challenge informer
- Make queuing handler generic and accept typed predicates
- Evaluate predicates on create/update/delete before enqueueing objects
- Add typed event handling, tombstone delete support, and logging

```release-note
NONE
```

inteon commented on pkg/controller/acmechallenges/controller.go:
```diff
@@ -135,7 +140,34 @@ func (c *controller) Register(ctx *controllerpkg.Context) (workqueue.TypedRateLi
-	if _, err := challengeInformer.Informer().AddEventHandler(&controllerpkg.QueuingEventHandler{Queue: c.queue}); err != nil {
+	if _, err := challengeInformer.Informer().AddEventHandler(&controllerpkg.QueuingEventHandler{
+		Queue: c.queue,
+		Predicates: []predicate.TypedPredicate[client.Object]{
```
@wallrj-cyberark maybe the queue and predicates can use the *cmacme.Challenge type instead of client.Object?
We can do the casting in the event handler.

inteon committed:
Event handler: add support for predicate based filtering

Co-authored-by: Richard Wall <richard.wall@cyberark.com>
Signed-off-by: Tim Ramlot <42113979+inteon@users.noreply.github.com>

inteon committed:
Ignore Challenge Status.Reason-only updates

- Use typed predicate in acmechallenges to ignore Status.Reason-only
- Remove stabilizeSolverErrorMessage and set Status.Reason to err.Error()

Signed-off-by: Richard Wall <richard.wall@cyberark.com>

inteon committed:
Make event handlers correctly typed, using generics

Signed-off-by: Tim Ramlot <42113979+inteon@users.noreply.github.com>

inteon committed:
Use resource version instead of deepequal when possible

Signed-off-by: Tim Ramlot <42113979+inteon@users.noreply.github.com>
